# Chapter 3 — Dictionaries and Sets

In [1]:
import sys

print(sys.version)

3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]


In [2]:
from collections import abc

my_dict = {}
isinstance(my_dict, abc.Mapping)

True

In [3]:
isinstance(my_dict, abc.MutableMapping)

True

In [4]:
tt = (1, 2, (30, 40))
hash(tt)

-3907003130834322577

In [5]:
tl = (1, 2, [30, 40])
try:
    hash(tl)
except TypeError as e:
    print(e)

unhashable type: 'list'


In [6]:
tf = (1, 2, frozenset([30, 40]))
hash(tf)

5149391500123939311

In [7]:
a = dict(one=1, two=2, three=3)
b = {'three': 3, 'two': 2, 'one': 1}
c = dict([('two', 2), ('one', 1), ('three', 3)])
d = dict(zip(['one', 'two', 'three'], [1, 2, 3]))
e = dict({'three': 3, 'one': 1, 'two': 2})
a == b == c == d == e

True

In [8]:
a

{'one': 1, 'two': 2, 'three': 3}

In [9]:
list(a.keys())

['one', 'two', 'three']

In [10]:
c

{'two': 2, 'one': 1, 'three': 3}

In [11]:
c.popitem()

('three', 3)

In [12]:
c

{'two': 2, 'one': 1}

## Modern dict Syntax

#### Example 3-1. Examples of dict comprehensions

In [13]:
dial_codes = [  # <1>
    (880, 'Bangladesh'),
    (55, 'Brazil'),
    (86, 'China'),
    (91, 'India'),
    (62, 'Indonesia'),
    (81, 'Japan'),
    (234, 'Nigeria'),
    (92, 'Pakistan'),
    (7, 'Russia'),
    (1, 'United States'),
]

In [14]:
country_dial = {country: code for code, country in dial_codes}  # <2>
country_dial

{'Bangladesh': 880,
 'Brazil': 55,
 'China': 86,
 'India': 91,
 'Indonesia': 62,
 'Japan': 81,
 'Nigeria': 234,
 'Pakistan': 92,
 'Russia': 7,
 'United States': 1}

In [15]:
{code: country.upper()  # <3>
 for country, code in sorted(country_dial.items())
 if code < 70}

{55: 'BRAZIL', 62: 'INDONESIA', 7: 'RUSSIA', 1: 'UNITED STATES'}

In [16]:
def dump(**kwargs):
    return kwargs


dump(**{'x': 1}, y=2, **{'z': 3})

{'x': 1, 'y': 2, 'z': 3}

In [17]:
d1 = {'a': 1, 'b': 3}
d2 = {'a': 2, 'b': 4, 'c': 6}
{'a': 2, 'b': 4, 'c': 6}

{'a': 2, 'b': 4, 'c': 6}

In [18]:
d1

{'a': 1, 'b': 3}

In [19]:
d1 |= d2
d1

{'a': 2, 'b': 4, 'c': 6}

## Pattern Matching with Mappings

#### Example 3-2. `creator.py`: `get_creators()` extracts names of creators from media records

**Note**: Python 3.10 is required

In [20]:
def get_creators(record: dict) -> list:
    match record:
        case {'type': 'book', 'api': 2, 'authors': [*names]}:  # <1>
            return names
        case {'type': 'book', 'api': 1, 'author': name}:  # <2>
            return [name]
        case {'type': 'book'}:  # <3>
            raise ValueError(f"Invalid 'book' record: {record!r}")
        case {'type': 'movie', 'director': name}:  # <4>
            return [name]
        case _:  # <5>
            raise ValueError(f'Invalid record: {record!r}')

In [21]:
b1 = dict(api=1, author='Douglas Hofstadter',
          type='book', title='Gödel, Escher, Bach')
get_creators(b1)

['Douglas Hofstadter']

In [22]:
from collections import OrderedDict

b2 = OrderedDict(api=2, type='book',
                 title='Python in a Nutshell',
                 authors='Martelli Ravenscroft Holden'.split())

In [23]:
get_creators(b2)

['Martelli', 'Ravenscroft', 'Holden']

In [24]:
get_creators({'type': 'book', 'pages': 770})

ValueError: Invalid 'book' record: {'type': 'book', 'pages': 770}

In [25]:
get_creators('Spam, spam, spam')

ValueError: Invalid record: 'Spam, spam, spam'

In [26]:
food = dict(category='ice cream', flavor='vanilla', cost=199)
match food:
    case {'category': 'ice cream', **details}:
        print(f'Ice cream details: {details}')

Ice cream details: {'flavor': 'vanilla', 'cost': 199}


## Standard API of Mapping Types

### What is Hashable

In [27]:
my_dict = {}
isinstance(my_dict, abc.Mapping)

True

In [28]:
isinstance(my_dict, abc.MutableMapping)

True

In [29]:
tt = (1, 2, (30, 40))
hash(tt)

-3907003130834322577

In [30]:
tl = (1, 2, [30, 40])
hash(tl)

TypeError: unhashable type: 'list'

In [31]:
tf = (1, 2, frozenset([30, 40]))
hash(tf)

5149391500123939311

### Inserting or Updating Mutable Values

#### Example 3-4. [`index0.py`](index0.py): Processing the "Zen of Python"; each line shows a word and a list of occurences coded as pairs

In [32]:
# !python3 index0.py zen.txt

#### Example 3-5. [`index.py`](index.py) uses `dict.setdefault` to fetch and update a list of word occurrences from the index in a single line; contrast with Example 3-4.

In [33]:
# python3 index.py zen.txt

## Automatic handling of Missing Keys

### `defaultdict`: Another take on missing keys

#### Example 3-6. [`index_default.py`](index_default.py): using defaultdict instead of the setdefault method

In [34]:
# !python index_default.py zen.txt

### The `__missing__` method

#### Example 3-7. When searching for a nonstring key, [`StrKeyDict0`](strkeydict0.py) converts it to `str` when it is not found

In [35]:
from strkeydict0 import StrKeyDict0

d = StrKeyDict0([('2', 'two'), ('4', 'four')])
d['2']

'two'

In [36]:
d[4]

'four'

In [37]:
d[1]

KeyError: '1'

In [38]:
d.get('2')

'two'

In [39]:
d.get(4)

'four'

In [40]:
d.get(1, 'N/A')

'N/A'

In [41]:
2 in d

True

In [42]:
1 in d

False

## Variations of `dict`

### `collections.ChainMap`

In [43]:
d1 = dict(a=1, b=3)
d2 = dict(a=2, b=4, c=6)
from collections import ChainMap
chain = ChainMap(d1, d2)
chain['a']

1

In [44]:
chain['c']

6

In [45]:
chain['c'] = -1
d1

{'a': 1, 'b': 3, 'c': -1}

In [46]:
d2

{'a': 2, 'b': 4, 'c': 6}

### `collections.Counter`

In [47]:
import collections
ct = collections.Counter('abracadabra')
ct

Counter({'a': 5, 'b': 2, 'r': 2, 'c': 1, 'd': 1})

In [48]:
ct.update('aaaaazzz')
ct

Counter({'a': 10, 'b': 2, 'r': 2, 'c': 1, 'd': 1, 'z': 3})

In [49]:
ct.most_common(3)

[('a', 10), ('z', 3), ('b', 2)]

## Immutable Mappings

#### Example 3-10. `MappingProxyType` builds a read-only `mappingproxy` instance from a `dict`

In [50]:
from types import MappingProxyType
d = {1: 'A'}
d_proxy = MappingProxyType(d)
d_proxy

mappingproxy({1: 'A'})

In [51]:
d_proxy[1]

'A'

In [52]:
d_proxy[2] = 'x'

TypeError: 'mappingproxy' object does not support item assignment

In [53]:
d[2] = 'B'
d_proxy

mappingproxy({1: 'A', 2: 'B'})

In [54]:
d_proxy[2]

'B'

## Dictionary Views

#### Example 3-11. The `.values()` method returns a view of the values in a `dict`

In [55]:
d = dict(a=10, b=20, c=30)
values = d.values()
values

dict_values([10, 20, 30])

In [56]:
len(values)

3

In [57]:
list(values)

[10, 20, 30]

In [58]:
reversed(values)

In [59]:
values[0]

TypeError: 'dict_values' object is not subscriptable

In [60]:
d['z'] = 99
d

{'a': 10, 'b': 20, 'c': 30, 'z': 99}

In [61]:
values

dict_values([10, 20, 30, 99])

In [62]:
values_class = type({}.values())
v = values_class()

TypeError: cannot create 'dict_values' instances

## Set Theory

In [63]:
l = ['spam', 'spam', 'eggs', 'spam', 'bacon', 'eggs']
set(l)

{'bacon', 'eggs', 'spam'}

In [64]:
list(set(l))

['bacon', 'spam', 'eggs']

In [65]:
# Preserve the order of the first occurence
dict.fromkeys(l).keys()

dict_keys(['spam', 'eggs', 'bacon'])

In [66]:
list(dict.fromkeys(l).keys())

['spam', 'eggs', 'bacon']

### Set Literals

In [67]:
s = {1}
type(s)

set

In [68]:
s

{1}

In [69]:
s.pop()

1

In [70]:
s

set()

In [71]:
frozenset(range(10))

frozenset({0, 1, 2, 3, 4, 5, 6, 7, 8, 9})

### Set Comprehensions

#### Example 3-15. Build a set of Latin-1 characters that have the word “SIGN” in their Unicode names

In [72]:
from unicodedata import name
{chr(i) for i in range(32, 256) if 'SIGN' in name(chr(i),'')}

{'#',
 '$',
 '%',
 '+',
 '<',
 '=',
 '>',
 '¢',
 '£',
 '¤',
 '¥',
 '§',
 '©',
 '¬',
 '®',
 '°',
 '±',
 'µ',
 '¶',
 '×',
 '÷'}

## Set Operations on dict Views

In [73]:
d1 = dict(a=1, b=2, c=3, d=4)
d2 = dict(b=20, d=40, e=50)
d1.keys() & d2.keys()

{'b', 'd'}

In [74]:
s = {'a', 'e', 'i'}
d1.keys() & s

{'a'}

In [75]:
d1.keys() | s

{'a', 'b', 'c', 'd', 'e', 'i'}